pkg/common/commands/resetoffset/refmappers/subscription.go

```diff
@@ -0,0 +1,155 @@
+type SubscriptionConsumerGroupIdMapper func(*messagingv1.Subscription) (string, error)
+
+// Verify The Subscription ResetOffsetRefMapper Implements The Interface
+var _ ResetOffsetRefMapper = &SubscriptionRefMapper{}
+
+// SubscriptionRefMapper implements the ResetOffsetRefMapper for Knative Subscriptions
+type SubscriptionRefMapper struct {
+	logger             *zap.Logger
+	subscriptionLister messaginglisters.SubscriptionLister
+	topicNameMapper    SubscriptionTopicNameMapper
+	groupIdMapper      SubscriptionConsumerGroupIdMapper
+}
+
+// NewSubscriptionRefMapper returns an initialized ResetOffsetSubscriptionRefMapper
+func NewSubscriptionRefMapper(ctx context.Context,
+	topicNameMapper SubscriptionTopicNameMapper,
+	groupIdMapper SubscriptionConsumerGroupIdMapper) *SubscriptionRefMapper {
+
+	// Get The Logger From Context
+	logger := logging.FromContext(ctx).Desugar()
+
+	// Get The Subscription Informer From Context (Context Must Have Injected Informers From SharedMain())
+	subscriptionInformer := subscriptioninformers.Get(ctx)
+
+	// Return An Initialized ResetOffsetSubscriptionRefMapper
+	return &SubscriptionRefMapper{
+		logger:             logger,
+		subscriptionLister: subscriptionInformer.Lister(),
+		topicNameMapper:    topicNameMapper,
+		groupIdMapper:      groupIdMapper,
+	}
+}
+
+// MapRef implements the ResetOffsetRefMapper interface for Subscription references. It will return an
+// error in all cases other than successfully mapping the ResetOffset.Spec.Ref to a Kafka Topic / Group.
+func (m *SubscriptionRefMapper) MapRef(resetOffset *kafkav1alpha1.ResetOffset) (string, string, error) {
+
+	// Validate The ResetOffset
+	if resetOffset == nil {
+		m.logger.Warn("Received nil ResetOffset argument")
+		return "", "", fmt.Errorf("unable to map nil ResetOffset")
+	}
+
+	// Get The ResetOffset Ref From Spec & Enhance Logger
+	ref := resetOffset.Spec.Ref
+	logger := m.logger.With(zap.Any("Ref", ref))
+
+	// Validate The Reference
+	if !strings.HasPrefix(ref.APIVersion, messaging.GroupName) ||
+		ref.Kind != "Subscription" ||
+		len(ref.Name) <= 0 {
+		m.logger.Warn("Received ResetOffset with invalid Subscription reference")
+		return "", "", fmt.Errorf("received ResetOffset with invalid Subscription reference: %v", ref)
+	}
+
+	// Default Optional Ref.Namespace If Not Provided
+	refNamespace := ref.Namespace
+	if len(refNamespace) <= 0 {
+		refNamespace = resetOffset.Namespace
+	}
+
+	// Attempt To Get The Specified Subscription
+	subscription, err := m.subscriptionLister.Subscriptions(refNamespace).Get(ref.Name)
+	if err != nil {
+		logger.Error("Failed to get Subscription referenced by ResetOffset", zap.Error(err))
+		return "", "", fmt.Errorf("failed to get Subscription referenced by ResetOffset.Spec.Ref '%v': %v", ref, err)
+	}
+	if subscription == nil {
+		logger.Info("No Subscription found for ResetOffset reference")
```

Comment:
I think `Info` is appropriate here but good question.  The reason is that it's not a failure/error as nothing unexpected has happened - only that the user has created an invalid ResetOffset CRD (pointing to non-existent Subscription) and we're just logging that fact.  If it was something the code was creating and it should always be there then a Warning might make more sense.   sound ok?

Reply:
It does, especially since the code that calls MapRef will log its own error when this function returns an error.  Originally I was thinking "the admin will be looking for an error in the logs as to why the ResetOffset CR didn't work," but there will still be a "Failed to map ResetOffset.Spec.Ref to Kafka Topic name and ConsumerGroup ID" error in that case.  So I agree.